codec/src/main/java/io/netty/handler/codec/compression/ByteBufChecksum.java

```diff
@@ -33,43 +32,34 @@
  * byte array ({@link ByteBuf#hasArray()} is {@code true}) or not.
  */
 abstract class ByteBufChecksum implements Checksum {
-    private static final Method ADLER32_UPDATE_METHOD;
-    private static final Method CRC32_UPDATE_METHOD;
-
-    static {
-        // See if we can use fast-path when using ByteBuf that is not heap based as Adler32 and CRC32 added support
-        // for update(ByteBuffer) in JDK8.
-        ADLER32_UPDATE_METHOD = updateByteBuffer(new Adler32());
-        CRC32_UPDATE_METHOD = updateByteBuffer(new CRC32());
-    }
 
     private final ByteProcessor updateProcessor = value -> {
         update(value);
         return true;
     };
 
-    private static Method updateByteBuffer(Checksum checksum) {
-        try {
-            Method method = checksum.getClass().getDeclaredMethod("update", ByteBuffer.class);
-            method.invoke(checksum, ByteBuffer.allocate(1));
-            return method;
-        } catch (Throwable ignore) {
-            return null;
-        }
-    }
-
     static ByteBufChecksum wrapChecksum(Checksum checksum) {
         requireNonNull(checksum, "checksum");
         if (checksum instanceof ByteBufChecksum) {
             return (ByteBufChecksum) checksum;
         }
-        if (checksum instanceof Adler32 && ADLER32_UPDATE_METHOD != null) {
-            return new ReflectiveByteBufChecksum(checksum, ADLER32_UPDATE_METHOD);
+        if (checksum instanceof Adler32) {
+            return new OptimizedByteBufChecksum<Adler32>((Adler32) checksum) {
+                @Override
+                public void update(ByteBuffer b) {
+                    checksum.update(b);
+                }
+            };
         }
-        if (checksum instanceof CRC32 && CRC32_UPDATE_METHOD != null) {
-            return new ReflectiveByteBufChecksum(checksum, CRC32_UPDATE_METHOD);
+        if (checksum instanceof CRC32) {
+            return new OptimizedByteBufChecksum<CRC32>((CRC32) checksum) {
```

Comment:
I would create a  `CRC32 castedChecksum = (CRC32) checksum` to be used on 
```java
                @Override
                public void update(ByteBuffer b) {
                    castedChecksum.update(b);
                }
```

Reply:
Same for `Adler32`: IIRC captured final are `true finals` 4 the JVM, so I would expect `checksum.update(b);` to be monomorphic, but to reduce the chance that the JVM won't be able to do that, this change would help...

Reply:
I would prefer to keep it as it is as its cleaner imho and I am not convinced it matters.

Reply:
Yep,. probably it won't worth doing it